posts/how-to-troubleshoot-java-NegativeArraySizeException.md

```diff
@@ -0,0 +1,298 @@
+
+You may encounter the exception as follows.
+
+<Highlighter
+input={`
+ 
+         Exception in thread "main" java.lang.NegativeArraySizeException: -2
+         at example.main(example.java:6) 
+   `}
+language='bash'
+/>
+
+This error can arise either when the user inputs a negative integer for the array size or the array size is the result of a mathematical expression(or an arithmetic operation). Let’s consider both cases.
+
+## Reproducing the error — user input
+
+This error arises when the array length is initialized with a negative input by the user.
+
+The following program accepts the size of an array and the array elements from the user and then returns a sorted array.
+
+<Highlighter
+input={`import java.util.Arrays;
+import java.io.*; 
+ 
+    class sorting
+      {
+        public static void main(String [] args) throws IOException
+        {
+          System.out.println(
+            "The sorted array is " + Arrays.toString(
+              input())
+          );
+        }
+            static int[] input()throws IOException // A function to accept input from the user
+              {
+                BufferedReader br=new BufferedReader(
+                  new InputStreamReader(System.in)
+                );
+                System.out.println("Enter the array size");
+                int size=Integer.parseInt(
+                  br.readLine()
+                ); // Accepting the size of the array from the user
+                int[] arr= new int[size];
+                
+                System.out.println(
+                  "Now Enter the elements one by one that are to be sorted"
+                );
+                for(int i=0;i<size;i++)
+                  { arr[i]=Integer.parseInt(br.readLine()); }
+                Arrays.sort(arr);
+              return arr;
+              }
+      }
+ `}
+language='java'
+/>
+
+This is a program that accepts an array from the user and will sort and output the array. Now, let's enter a negative input for the array size, this produces the **NegativeArraySizeException** as below.
+ 
+ <Highlighter
+input={`Enter the array size
+-5
+Exception in thread "main" java.lang.NegativeArraySizeException: -5
+at sorting.input(sorting.java:14)
+at sorting.main(sorting.java:7)`}
+language='bash'
+/>
+
+This program has thrown an error due to input from the user not having been anticipated by the program. 
+
+## Reproducing the error — arithmetic input
+
+This error arises when the array size is initialized with an arithmetic result. Let's consider an example to understand.
+
+We'll write a program to store the speeds of a body in the uniform intervals of 2M from the starting to the ending positions during its journey in either direction of `X` and `Y` axes.
+
+<Highlighter
+input={`import java.util.Arrays;
```

Comment:
I'm not getting the error when I run this code as written, here is my output:

```
Enter the starting position and ending position
7
2
Now Enter the speeds in m/s one after the other
The speeds in m/s are
 []
```

Reply:
@ltagliaferri , Yes that is the expected output, I considered this particular case while writing the logic, the arithmetic from **line 110-112** results in zero(the size variable is initialized with zero) for this particular input, actually to every input that has a difference of -5 has a similar output(like [6,1] , [5,0] , [4,-1]....). 

The easy fix to this is using the absolute function from the Math class, so the same code that uses **abs()** is the solution to this problem, like the code from **line 207-209**. I think we should actually show this case too and explain that "the size is initialized with zero for certain inputs although the exception here is arithmetic but not the negative array size exception" or we might choose to not show the case but we can simply add a line to **line 125** saying that the above code logic can also result in an arithmetic error which is usually not thrown by the compiler for certain inputs such as (7, 2) , (-10, -5), etc because the array length is initialized with zero in such cases.

Reply:
Ok, I see, we just need more context then. Thanks!